pkg/controller/service/service_controller.go

```diff
@@ -447,13 +329,21 @@ func (s *ServiceController) needsUpdate(oldService *v1.Service, newService *v1.S
 			oldService.Spec.Type, newService.Spec.Type)
 		return true
 	}
-
+	if !hasFinalizer(newService) && newService.DeletionTimestamp == nil {
+		// We only need to update the service if it's missing the finalizer
+		// and not marked for deletion. If it is marked and missing that means it
+		// was successfully removed already.
+		//
+		// This should only occur the first time this controller is deployed and existing
```

Comment:
It'll be the same thing that happens now; it could leak if certain things happen like say the controller manager restarts during the delete.

Reply:
Right, even today if the user deletes the service while the controller is restarting than its LB will be leaked. So that remains same.

But once we have finalizers setup such that all existing services get a finalizer, than controller will not leak LBs even across restarts.

sgtm.